### PR TITLE
core: SpiffeUtil API for extracting Spiffe URI and loading TrustBundles

### PR DESCRIPTION
**Pull request by erm-g**
Additional API for SpiffeUtil:

- extract Spiffe URI from certificate chain
- load Spiffe Trust Bundle from filesystem [json spec](https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE_Trust_Domain_and_Bundle.md) [JWK spec](https://github.com/spiffe/spiffe/blob/main/standards/X509-SVID.md#61-publishing-spiffe-bundle-elements)

### CODE REVIEW

File: core/src/main/java/io/grpc/internal/SpiffeUtil.java [hunk 1]
```diff
@@ -96,6 +126,143 @@ private static void validatePathSegment(String pathSegment) {
...
+            + "domain %s.", rawCerts.size(), trustDomainName));
+        break;
+      }
+      InputStream stream = new ByteArrayInputStream((CERTIFICATE_PREFIX + System.lineSeparator()
+          + rawCerts.get(0) + System.lineSeparator() + CERTIFICATE_SUFFIX)
+          .getBytes(StandardCharsets.UTF_8));
+      try {
+        Collection<? extends Certificate> certs = CertificateFactory.getInstance("X509")
+            .generateCertificates(stream);
+        result.add(certs.toArray(new X509Certificate[0])[0]);
```
Comment (matthewstevenson88) [inline core/src/main/java/io/grpc/internal/SpiffeUtil.java:257]: Should we check that `certs.size()` is exactly 1?

File: core/src/main/java/io/grpc/internal/SpiffeUtil.java [hunk 2]
```diff
@@ -119,4 +286,34 @@ public String getPath() {
...
+   */
+  public static final class SpiffeBundle {
+
+    private final ImmutableMap<String, Long> sequenceNumbers;
+
+    private final ImmutableMap<String, ImmutableList<X509Certificate>> bundleMap;
+
+    private SpiffeBundle(Map<String, Long> sequenceNumbers,
+        Map<String, List<X509Certificate>> trustDomainMap) {
+      this.sequenceNumbers = ImmutableMap.copyOf(sequenceNumbers);
```
Comment (matthewstevenson88) [inline core/src/main/java/io/grpc/internal/SpiffeUtil.java:302]: micro-nit: "this." is redundant here and can be removed.

File: core/src/main/java/io/grpc/internal/SpiffeUtil.java [hunk 3]
```diff
@@ -119,4 +286,34 @@ public String getPath() {
...
+      ImmutableMap.Builder<String, ImmutableList<X509Certificate>> builder = ImmutableMap.builder();
+      for (Map.Entry<String, List<X509Certificate>> entry : trustDomainMap.entrySet()) {
+        builder.put(entry.getKey(), ImmutableList.copyOf(entry.getValue()));
+      }
+      this.bundleMap = builder.build();
```
Comment (matthewstevenson88) [inline core/src/main/java/io/grpc/internal/SpiffeUtil.java:307]: micro-nit: "this." is redundant here and can be removed.

File: core/src/main/java/io/grpc/internal/SpiffeUtil.java [hunk 4]
```diff
@@ -96,6 +126,143 @@ private static void validatePathSegment(String pathSegment) {
...
+    }
+    if (uri == null) {
+      return Optional.absent();
+    }
+    return Optional.of(parse(uri));
+  }
+
+  /**
+   * Loads a SPIFFE trust bundle from a file, parsing it from the JSON format.
+   * In case of success, returns trust domains, their associated sequence numbers, and X.509
```
Comment (matthewstevenson88) [inline core/src/main/java/io/grpc/internal/SpiffeUtil.java:162]: Two minor presentation comments:
1. Rather than redefining a SPIFFE bundle, I think it suffices to say we return a SPIFFE bundle. :)
2. Please clarify that, if any "key" in the trust map is deemed to be invalid or unsupported", we log and drop it from the bundle, but do not fail overall.

File: core/src/main/java/io/grpc/internal/SpiffeUtil.java [hunk 5]
```diff
@@ -119,4 +286,34 @@ public String getPath() {
     }
   }
 
+  /**
+   * Represents a Trust Bundle inspired by SPIFFE Bundle Format standard. Only trust domain's
```
Comment (matthewstevenson88) [inline core/src/main/java/io/grpc/internal/SpiffeUtil.java:290]: Suggested edit for clarity: Represents a SPIFFE trust bundle; that is, a map from trust domain to set of trusted certificates.

File: core/src/test/java/io/grpc/internal/SpiffeUtilTest.java [hunk 6]
```diff
@@ -193,4 +207,119 @@ public void spiffePathFormatTest() {
           + "underscores ([a-zA-Z0-9.-_])", iae.getMessage());
     }
   }
+
+  public static class CertificateApiTest {
```
Comment (matthewstevenson88) [inline core/src/test/java/io/grpc/internal/SpiffeUtilTest.java:211]: nit: "CertificateApiTest" is very vague - can we say "ExtractSpiffeIdTest" instead, so that the test name is closely tied to the API we are testing?

I also see that we're grouping this together with the "load SPIFFE bundle APIs" - should we separate them, for consistency with what we did for parsing?

File: core/src/test/java/io/grpc/internal/SpiffeUtilTest.java [hunk 7]
```diff
@@ -193,4 +207,119 @@ public void spiffePathFormatTest() {
...
+      spiffeCert = CertificateUtils.getX509Certificates(TlsTesting.loadCert(SPIFFE_PEM_FILE));
+      multipleUriSanCert = CertificateUtils.getX509Certificates(TlsTesting
+          .loadCert(MULTI_URI_SAN_PEM_FILE));
+      serverCert0 = CertificateUtils.getX509Certificates(TlsTesting.loadCert(SERVER_0_PEM_FILE));
+    }
+
+    @Test
+    public void extractSpiffeIdSuccessTest() throws CertificateParsingException {
+      Optional<SpiffeId> spiffeId = SpiffeUtil.extractSpiffeId(spiffeCert);
+      assertEquals("foo.bar.com", spiffeId.get().getTrustDomain());
```
Comment (matthewstevenson88) [inline core/src/test/java/io/grpc/internal/SpiffeUtilTest.java:240]: nit: I'd suggest having an initial assertion `assertTrue(spiffeId.isPresent())` before checking the trust domain and path. This way it makes it very clear what the problem is if a buggy change is made to the `extractSpiffeId` API.

Similarly throughout.

File: core/src/test/java/io/grpc/internal/SpiffeUtilTest.java [hunk 8]
```diff
@@ -193,4 +207,119 @@ public void spiffePathFormatTest() {
...
+      assertEquals("/client/workload/1", spiffeId.get().getPath());
+    }
+
+    @Test
+    public void extractSpiffeIdFailureTest() throws CertificateParsingException {
+      Optional<SpiffeUtil.SpiffeId> spiffeId = SpiffeUtil.extractSpiffeId(serverCert0);
+      assertFalse(spiffeId.isPresent());
+      IllegalArgumentException iae = assertThrows(IllegalArgumentException.class, () -> SpiffeUtil
```
Comment (matthewstevenson88) [inline core/src/test/java/io/grpc/internal/SpiffeUtilTest.java:248]: optional: I generally prefer to avoid non-standard abbreviations in code and tests, so consider replacing "iae" with e.g. "exception". Similarly throughout. Non-blocking optional comment.

File: core/src/test/java/io/grpc/internal/SpiffeUtilTest.java [hunk 9]
```diff
@@ -193,4 +207,119 @@ public void spiffePathFormatTest() {
...
+          .extractSpiffeId(multipleUriSanCert));
+      assertEquals("Multiple URI SAN values found in the leaf cert.", iae.getMessage());
+
+    }
+
+    @Test
+    public void extractSpiffeIdFromChainTest() throws CertificateParsingException {
+      X509Certificate[] leafWithSpiffeChain = new X509Certificate[]{spiffeCert[0], serverCert0[0]};
+      assertTrue(SpiffeUtil.extractSpiffeId(leafWithSpiffeChain).isPresent());
```
Comment (matthewstevenson88) [inline core/src/test/java/io/grpc/internal/SpiffeUtilTest.java:257]: Please also check the value of the trust domain and path, to guarantee that we have extract the "right" SPIFFE ID from the chain.

File: core/src/test/java/io/grpc/internal/SpiffeUtilTest.java [hunk 10]
```diff
@@ -193,4 +207,119 @@ public void spiffePathFormatTest() {
...
+      NullPointerException npe = assertThrows(NullPointerException.class, () -> SpiffeUtil
+          .extractSpiffeId(null));
+      assertEquals("certChain", npe.getMessage());
+      IllegalArgumentException iae = assertThrows(IllegalArgumentException.class, () -> SpiffeUtil
+          .extractSpiffeId(new X509Certificate[]{}));
+      assertEquals("CertChain can't be empty", iae.getMessage());
+    }
+
+    @Test
+    public void loadTrustBundleFromFileSuccessTest() throws IOException,
```
Comment (matthewstevenson88) [inline core/src/test/java/io/grpc/internal/SpiffeUtilTest.java:274]: nit: In unit tests, it's ok (and common) to write "throws Exception".

File: core/src/test/java/io/grpc/internal/SpiffeUtilTest.java [hunk 11]
```diff
@@ -193,4 +207,119 @@ public void spiffePathFormatTest() {
...
+      assertEquals(0, tb.getBundleMap().get("test.google.com").size());
+      assertEquals(0, tb.getBundleMap().get("test.google.com.au").size());
+      assertEquals(1, tb.getBundleMap().get("example.com").size());
+      assertEquals(2, tb.getBundleMap().get("test.example.com").size());
+      assertEquals("foo.bar.com", SpiffeUtil.extractSpiffeId(tb.getBundleMap().get("example.com")
+          .toArray(new X509Certificate[0])).get().getTrustDomain());
+    }
+
+    @Test
+    public void loadTrustBundleFromFileFailureTest() throws IOException, CertificateException {
```
Comment (matthewstevenson88) [inline core/src/test/java/io/grpc/internal/SpiffeUtilTest.java:294]: Can we either separate the different failure cases into different tests, or at the very least add comments separating them and giving a very brief description of the test case?

File: core/src/test/java/io/grpc/internal/SpiffeUtilTest.java (hunk 10 above)
Comment (matthewstevenson88) [inline core/src/test/java/io/grpc/internal/SpiffeUtilTest.java:274]: To simplify the readability and maintainability of the tests, consider the following pattern:

```

public void test() {
    SpiffeBundle spiffeBundle = ...;
    assertEquals(spiffeBundle.getSequenceNumbers(), expectedSequenceNumbers());
    assertEquals(spiffeBundle.getBundleMap(), expectedBundleMap());
}

private static ImmutableMap<String, Long> expectedSequenceNumbers() {
  return {{"google.com", 123L}, ...}
}

private static ImmutableMap<String, ImmutableList<X509Certificate>> expectedBundleMap() {
    // construct explicit expected bundle map
}

```

Comment (erm-g) [reply core/src/test/java/io/grpc/internal/SpiffeUtilTest.java:274]: Talking about 'method returning expected data' pattern - IMHO it's worth it if its used in >1 place, so imho inlining creation of the 'expected map' looks better here.
Talking about 'comparing collections instead of individual elements' pattern - it's definitely smth to consider but I borrowed the current style from https://github.com/grpc/grpc-java/blob/master/auth/src/test/java/io/grpc/auth/GoogleAuthLibraryCallCredentialsTest.java#L262

File: core/src/test/java/io/grpc/internal/SpiffeUtilTest.java (hunk 11 above)
Comment (erm-g) [reply core/src/test/java/io/grpc/internal/SpiffeUtilTest.java:294]: Failure cases are already grouped under 2 methods - failures with reading the file itself, and failures with JSON validation. I added comments but IMHO they're repeating the related exception messages

File: core/src/test/java/io/grpc/internal/SpiffeUtilTest.java (hunk 9 above)
Comment (erm-g) [reply core/src/test/java/io/grpc/internal/SpiffeUtilTest.java:257]: The values are already checked in another test - https://github.com/grpc/grpc-java/blob/821810f51b029ed4cef30eb421e3b4bd01c913ab/core/src/test/java/io/grpc/internal/SpiffeUtilTest.java#L240
So I'm trying to emphasize here the fact that we do extraction only from leafs

File: core/src/test/java/io/grpc/internal/SpiffeUtilTest.java (hunk 8 above)
Comment (erm-g) [reply core/src/test/java/io/grpc/internal/SpiffeUtilTest.java:248]: That's a standard way in this repo - https://github.com/grpc/grpc-java/blob/master/gcp-observability/src/test/java/io/grpc/gcp/observability/ObservabilityConfigImplTest.java#L223

File: core/src/test/java/io/grpc/internal/SpiffeUtilTest.java (hunk 6 above)
Comment (erm-g) [reply core/src/test/java/io/grpc/internal/SpiffeUtilTest.java:211]: Yeah, I couldn't find a good name here. So we test both extract and load APIs, and what unifies them is the fact that we do some operation with certs (extracting spiffe from them, or loading them in a structure).
If we separate them, then we'll need duplicate `setUp()` method to load the same certs for both. To avoid that we might introduce an extra parenting class for both, but it'll be the same question about how to name it.
WDYT?

File: core/src/main/java/io/grpc/internal/SpiffeUtil.java [hunk 12]
```diff
@@ -96,6 +126,143 @@ private static void validatePathSegment(String pathSegment) {
             + " ([a-zA-Z0-9.-_])");
   }
 
+  /**
+   * Returns the SPIFFE ID from the leaf certificate, if present.
+   *
+   * @param certChain certificate chain to extract SPIFFE ID from
+   */
+  public static Optional<SpiffeId> extractSpiffeId(X509Certificate[] certChain)
```
[the remark this answers is not in the document]
Comment (erm-g) [reply core/src/main/java/io/grpc/internal/SpiffeUtil.java:134]: Yeah, let's defer to Eric - IIUC the style checker also has some rules (at least I got violation few times) about it.

File: core/src/main/java/io/grpc/internal/SpiffeUtil.java (hunk 1 above)
Comment (erm-g) [reply core/src/main/java/io/grpc/internal/SpiffeUtil.java:257]: I believe we do it here - https://github.com/grpc/grpc-java/blob/821810f51b029ed4cef30eb421e3b4bd01c913ab/core/src/main/java/io/grpc/internal/SpiffeUtil.java#L246
IIUC since we manually add BEGIN and END CERT, then it's not possible to have a string (single element of `rawCerts` with multiple certs in it.

File: core/src/main/java/io/grpc/internal/SpiffeUtil.java (hunk 2 above)
Comment (erm-g) [reply core/src/main/java/io/grpc/internal/SpiffeUtil.java:302]: In this particular case it's mandatory - the arg name equals to class field name.

File: core/src/main/java/io/grpc/internal/SpiffeUtil.java (hunk 3 above)
Comment (erm-g) [reply core/src/main/java/io/grpc/internal/SpiffeUtil.java:307]: Yes, it's redundant but in this repo in constructors it's widely used - https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/internal/AtomicBackoff.java#L38

File: core/src/main/java/io/grpc/internal/SpiffeUtil.java (hunk 1 above)
Comment (matthewstevenson88) [reply core/src/main/java/io/grpc/internal/SpiffeUtil.java:249]: But could it be empty if `generateCertificates` fails somehow?

File: core/src/main/java/io/grpc/internal/SpiffeUtil.java (hunk 2 above)
Comment (matthewstevenson88) [reply core/src/main/java/io/grpc/internal/SpiffeUtil.java:294]: Facepalm, sorry you're totally correct. :)

File: core/src/test/java/io/grpc/internal/SpiffeUtilTest.java (hunk 6 above)
Comment (matthewstevenson88) [reply core/src/test/java/io/grpc/internal/SpiffeUtilTest.java:207]: Ack ok, fine leaving it.

File: core/src/test/java/io/grpc/internal/SpiffeUtilTest.java (hunk 9 above)
Comment (matthewstevenson88) [reply core/src/test/java/io/grpc/internal/SpiffeUtilTest.java:258]: But don't we need to make sure that we're grabbing the SPIFFE ID from the leaf and not from some intermediate?

Comment (erm-g) [reply core/src/test/java/io/grpc/internal/SpiffeUtilTest.java:258]: Ah, got it. I use the fact that serverCert0 doesn't have SPIFFE ID but spiffeCert does. That's not clear from the test - added the comment.

File: core/src/main/java/io/grpc/internal/SpiffeUtil.java (hunk 1 above)
Comment (erm-g) [reply core/src/main/java/io/grpc/internal/SpiffeUtil.java:249]: Yeah that's a tricky question. I couldn't create a test to make it return empty collection without failing (knowing the previous check of `rawCerts`), **but** the Javadoc for this method doesn't really guarantee that - https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/security/cert/CertificateFactory.html#generateCertificates(java.io.InputStream)
Keeping in mind that other JDKs might implement it slightly different, I think it's worth to add such check (I used `checkArgument` in that case, maybe some `assert` is more appropriate).

**Review by matthewstevenson88 — APPROVED** [review]

File: core/src/main/java/io/grpc/internal/JsonParser.java [hunk 13]
```diff
@@ -56,13 +56,32 @@ public static Object parse(String raw) throws IOException {
     }
   }
 
-  private static Object parseRecursive(JsonReader jr) throws IOException {
+  /**
+   * Parses a json string, returning either a {@code Map<String, ?>}, {@code List<?>},
+   * {@code String}, {@code Double}, {@code Boolean}, or {@code null}. Fails if duplicate names
+   * found.
+   */
+  public static Object parseNoDuplicates(String raw) throws IOException {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/JsonParser.java:64]: I thought we were going to change the code to do this all the time. C was doing it already, so it is already reasonably safe, and if you have duplicates you really don't know what result you'll get.

If you are concern about this causing problems, then we could have a flag set from an environment variable. We default it to "fail on duplicate" but someone can change it if it is a problem. After a few releases, we drop the old way.

File: core/src/main/java/io/grpc/internal/JsonParser.java [hunk 14]
```diff
@@ -56,13 +56,32 @@ public static Object parse(String raw) throws IOException {
...
+    } finally {
+      try {
+        jr.close();
+      } catch (IOException e) {
+        logger.log(Level.WARNING, "Failed to close", e);
+      }
+    }
+  }
+
+  private static Object parseRecursive(JsonReader jr, Boolean... failOnDuplicates)
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/JsonParser.java:77]: Why is this using varargs?

File: core/src/main/java/io/grpc/internal/SpiffeUtil.java [hunk 15]
```diff
@@ -96,6 +123,140 @@ private static void validatePathSegment(String pathSegment) {
...
+      List<String> rawCerts = JsonUtil.getListOfStrings(keyNode, "x5c");
+      if (rawCerts == null) {
+        break;
+      }
+      if (rawCerts.size() != 1) {
+        throw new IllegalArgumentException(String.format("Exactly 1 certificate is expected, but "
+            + "%s found. Certificate loading for trust domain '%s' failed.", rawCerts.size(),
+            trustDomainName));
+      }
+      InputStream stream = new ByteArrayInputStream((CERTIFICATE_PREFIX + System.lineSeparator()
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/SpiffeUtil.java:243]: Why are we using the system line separator? Isn't the format here cross-platform?

Feel free to use literals here. It is very clear as:
```java
"-----BEGIN CERTIFICATE-----\n"
+ rawCerts.get(0) + "\n"
+ "-----END CERTIFICATE-----"
```

File: core/src/main/java/io/grpc/internal/SpiffeUtil.java [hunk 16]
```diff
@@ -96,6 +123,140 @@ private static void validatePathSegment(String pathSegment) {
...
+
+  private static void checkJwkEntry(Map<String, ?> jwkNode, String trustDomainName) {
+    String kty = JsonUtil.getString(jwkNode, "kty");
+    if (kty == null || !kty.equals(KTY_PARAMETER_VALUE)) {
+      throw new IllegalArgumentException(String.format("'kty' parameter must be '%s' but '%s' "
+              + "found. Certificate loading for trust domain '%s' failed.", KTY_PARAMETER_VALUE,
+          kty, trustDomainName));
+    }
+    String kid = JsonUtil.getString(jwkNode, "kid");
+    if (kid != null && !kid.equals("")) {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/SpiffeUtil.java:216]: Seems we shouldn't allow empty string. "Additionally, the kid parameter of each JWK entry MUST NOT be set." means not present at all. This would be better as `if (jwkNode.containsKey("kid")) throw new ...` as right now there's actually an assertion that the value is a String if it is present, when all you want is for it to not be present.

File: core/src/main/java/io/grpc/internal/SpiffeUtil.java [hunk 17]
```diff
@@ -96,6 +123,140 @@ private static void validatePathSegment(String pathSegment) {
...
+    Object jsonObject = JsonParser.parseNoDuplicates(json);
+    if (!(jsonObject instanceof Map)) {
+      throw new IllegalArgumentException(
+          "SPIFFE Trust Bundle should be a JSON object. Found: "
+              + (jsonObject == null ? null : jsonObject.getClass()));
+    }
+    @SuppressWarnings("unchecked")
+    Map<String, ?> root = (Map<String, ?>)jsonObject;
+    Map<String, ?> trustDomainsNode = JsonUtil.getObject(root, "trust_domains");
+    checkArgument(checkNotNull(trustDomainsNode,
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/SpiffeUtil.java:202]: Consider splitting this line in two.

File: cronet/build.gradle [hunk 18]
```diff
@@ -1,94 +1,94 @@
-plugins {
```
Comment (ejona86) [inline cronet/build.gradle:1]: Revert this change. Use `-PskipAndroid=true` or set `skipAndroid=true` is your gradle.properties.

File: core/src/main/java/io/grpc/internal/SpiffeUtil.java [hunk 19]
```diff
@@ -96,6 +123,140 @@ private static void validatePathSegment(String pathSegment) {
...
+   * @see <a href="https://github.com/spiffe/spiffe/blob/main/standards/X509-SVID.md#61-publishing-spiffe-bundle-elements">JWK entry format</a>
+   * @see <a href="https://datatracker.ietf.org/doc/html/rfc7517#appendix-B">x5c (certificate) parameter</a>
+   */
+  public static SpiffeBundle loadTrustBundleFromFile(String trustBundleFile) throws IOException {
+    Map<String, ?> trustDomainsNode = readTrustDomainsFromFile(trustBundleFile);
+    Map<String, List<X509Certificate>> trustBundleMap = new HashMap<>();
+    Map<String, Long> sequenceNumbers = new HashMap<>();
+    for (String trustDomainName : trustDomainsNode.keySet()) {
+      Map<String, ?> domainNode = JsonUtil.getObject(trustDomainsNode, trustDomainName);
+      if (domainNode == null || domainNode.size() == 0) {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/SpiffeUtil.java:174]: `null` is impossible, because that means the entry doesn't exist.

File: core/src/test/java/io/grpc/internal/SpiffeUtilTest.java [hunk 20]
```diff
@@ -193,4 +204,152 @@ public void spiffePathFormatTest() {
...
+              .toArray(new X509Certificate[0]));
+      assertTrue(spiffeId.isPresent());
+      assertEquals("foo.bar.com", spiffeId.get().getTrustDomain());
+    }
+
+    @Test
+    public void loadTrustBundleFromFileFailureTest() throws Exception {
+      // Check the exception if JSON root element is different from 'trust_domains'
+      NullPointerException npe = assertThrows(NullPointerException.class, () -> SpiffeUtil
+          .loadTrustBundleFromFile(Paths.get(ClassLoader.getSystemResource(TEST_DIRECTORY_PREFIX
```
Comment (ejona86) [inline core/src/test/java/io/grpc/internal/SpiffeUtilTest.java:300]: This is depending on specifics of the test environment that aren't guaranteed. In particular, with Blaze/Bazel this won't be the case. The necessary approach is to read the resource as a stream and copy it to a temporary file.

I was going to comment on the API that it would be better to pass an InputStream instead of a filename. Things like this become much easier that way (no need to make a temporary file).

File: core/src/main/java/io/grpc/internal/JsonParser.java (hunk 13 above)
Comment (erm-g) [reply core/src/main/java/io/grpc/internal/JsonParser.java:64]: Yeah, changing the default behavior would be ideal. Unfortunately I ran into an issue with
https://github.com/grpc/grpc-java/blob/master/gcp-observability/src/test/java/io/grpc/gcp/observability/ObservabilityConfigImplTest.java#L109
Does Observability conf allow duplicates?

File: core/src/test/java/io/grpc/internal/SpiffeUtilTest.java [hunk 21]
```diff
@@ -159,7 +170,8 @@ public void spiffeTrustDomainFormatTest() {
           SpiffeUtil.parse(longTrustDomain.toString()));
       assertEquals("Trust Domain maximum length is 255 characters", iae.getMessage());
 
-      StringBuilder longSpiffe = new StringBuilder(String.format("spiffe://mydomain%scom/", "%21"));
+      @SuppressWarnings("OrphanedFormatString")
+      StringBuilder longSpiffe = new StringBuilder("spiffe://mydomain%21com/");
```
Comment (erm-g) [inline core/src/test/java/io/grpc/internal/SpiffeUtilTest.java:174]: @ejona86 
The situation with that string is more interesting then we thought - actually they don't show warning for `https`. Filed a bug - https://github.com/google/error-prone/issues/4609

Comment (ejona86) [reply core/src/test/java/io/grpc/internal/SpiffeUtilTest.java:174]: Okay. Yeah, let's just suppress it for now. It produces harder-to-read code if trying to work around it in other ways.

File: core/src/main/java/io/grpc/internal/JsonParser.java [hunk 22]
```diff
@@ -56,7 +58,8 @@ public static Object parse(String raw) throws IOException {
     }
   }
 
-  private static Object parseRecursive(JsonReader jr) throws IOException {
+  private static Object parseRecursive(JsonReader jr)
+      throws IOException {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/JsonParser.java:62]: Unwrap. Ditto below.

File: core/src/main/java/io/grpc/internal/JsonParser.java [hunk 23]
```diff
@@ -76,11 +79,13 @@ private static Object parseRecursive(JsonReader jr) throws IOException {
...
   }
 
-  private static Map<String, ?> parseJsonObject(JsonReader jr) throws IOException {
+  private static Map<String, ?> parseJsonObject(JsonReader jr)
+      throws IOException {
     jr.beginObject();
     Map<String, Object> obj = new LinkedHashMap<>();
     while (jr.hasNext()) {
       String name = jr.nextName();
+      checkArgument(!obj.containsKey(name), "Duplicate key found: " + name);
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/JsonParser.java:88]: This allocates the error string even when there isn't an error. Use built-in formatting available to checkArgument.

File: core/src/main/java/io/grpc/internal/SpiffeUtil.java (hunk 19 above)
Comment (ejona86) [reply core/src/main/java/io/grpc/internal/SpiffeUtil.java]: Ditto for `kty` and `use`.

Comment (erm-g) [reply core/src/main/java/io/grpc/internal/SpiffeUtil.java]: I don't think I got it. For domainNode,
``` java
for (String trustDomainName : trustDomainsNode.keySet()) {
      Map<String, ?> domainNode = JsonUtil.getObject(trustDomainsNode, trustDomainName);
      if (domainNode == null || domainNode.size() == 0)
```
we iterate over the keys, and that's why `getObject` can't return null
For `kty` and `use`, we check if 

- jwkNode contains them
- the value matches the spec

so it is possible that for entries like 
```json
"keys": [
        {
          "x5c": "bytes"
        }
      ]
```
the values are null.
Am I missing something?

Comment (ejona86) [reply core/src/main/java/io/grpc/internal/SpiffeUtil.java]: I was thinking about how `getString()` does instanceof before returning. But yeah, if it the value is missing it will be null.

Comment (erm-g) [reply core/src/main/java/io/grpc/internal/SpiffeUtil.java]: I also realized that I didn't have any tests covering ClassCastException (there are few LOCs where it might happen). After adding one for  
```java
Long sequenceNumber = JsonUtil.getNumberAsLong(domainNode, "sequence_number");
```
I realized that I'm just piggybacking on `JsonUtil` exception pattern itself.
Yes, `JsonUtil` checks for key presence (and returns null if not found) before instanceof.

File: core/src/test/java/io/grpc/internal/SpiffeUtilTest.java [hunk 24]
```diff
@@ -193,4 +210,184 @@ public void spiffePathFormatTest() {
...
+    private static final String SPIFFE_TRUST_BUNDLE_WRONG_MULTI_CERTS =
+        "spiffebundle_wrong_multi_certs.json";
+    private static final String SPIFFE_TRUST_BUNDLE_DUPLICATES = "spiffebundle_duplicates.json";
+    private static final String SPIFFE_TRUST_BUNDLE_WRONG_ROOT = "spiffebundle_wrong_root.json";
+    private static final String SPIFFE_TRUST_BUNDLE_WRONG_SEQ = "spiffebundle_wrong_seq_type.json";
+    private static final String DOMAIN_ERROR_MESSAGE =
+        " Certificate loading for trust domain 'google.com' failed.";
+
+
+    private final TemporaryFolder tempFolder = new TemporaryFolder();
```
Comment (ejona86) [inline core/src/test/java/io/grpc/internal/SpiffeUtilTest.java:235]: Make this `public` and add `@Rule`, then you won't need `@After`.

File: core/src/test/java/io/grpc/internal/SpiffeUtilTest.java [hunk 25]
```diff
@@ -193,4 +210,184 @@ public void spiffePathFormatTest() {
...
+      tempSpiffeTrustBundleDuplicates = copyFileToTmp(SPIFFE_TRUST_BUNDLE_DUPLICATES);
+      tempSpiffeTrustBundleWrongRoot = copyFileToTmp(SPIFFE_TRUST_BUNDLE_WRONG_ROOT);
+      tempSpiffeTrustBundleWrongSeq = copyFileToTmp(SPIFFE_TRUST_BUNDLE_WRONG_SEQ);
+    }
+
+    private String copyFileToTmp(String fileName) throws Exception {
+      InputStream resourceStream = SpiffeUtilTest.class.getClassLoader()
+          .getResourceAsStream(TEST_DIRECTORY_PREFIX + fileName);
+      Path tempFilePath = tempFolder.newFile(fileName).toPath();
+      Files.copy(resourceStream, tempFilePath, StandardCopyOption.REPLACE_EXISTING);
```
Comment (ejona86) [inline core/src/test/java/io/grpc/internal/SpiffeUtilTest.java:276]: This doesn't close the InputStream. Create `resourceStream` with try-with-resources.

File: core/src/test/java/io/grpc/internal/SpiffeUtilTest.java [hunk 26]
```diff
@@ -193,4 +210,184 @@ public void spiffePathFormatTest() {
...
+
+    @Before
+    public void setUp() throws Exception {
+      tempFolder.create();
+      spiffeCert = CertificateUtils.getX509Certificates(TlsTesting.loadCert(SPIFFE_PEM_FILE));
+      multipleUriSanCert = CertificateUtils.getX509Certificates(TlsTesting
+          .loadCert(MULTI_URI_SAN_PEM_FILE));
+      serverCert0 = CertificateUtils.getX509Certificates(TlsTesting.loadCert(SERVER_0_PEM_FILE));
+
+      tempSpiffeTrustBundle = copyFileToTmp(SPIFFE_TRUST_BUNDLE);
```
Comment (ejona86) [inline core/src/test/java/io/grpc/internal/SpiffeUtilTest.java:260]: You didn't want to inline these to where they are used? Since you are deleting the entire folder, you don't need to track each in a variable. And each of these is only used once, so at the very least make it a local variable.

```java
SpiffeBundle tb = SpiffeUtil.loadTrustBundleFromFile(tempSpiffeTrustBundle);
// vs
String tempSpiffeTrustBundle = copyFileToTmp(SPIFFE_TRUST_BUNDLE);
SpiffeBundle tb = SpiffeUtil.loadTrustBundleFromFile(tempSpiffeTrustBundle);
// or
SpiffeBundle tb = SpiffeUtil.loadTrustBundleFromFile(copyFileToTmp(SPIFFE_TRUST_BUNDLE));
```

**Review by ejona86 — APPROVED** [review]
